File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/coding/_code_executor_agent.py
from typing import List, Sequence

from autogen_core.base import CancellationToken
from autogen_core.components.code_executor import CodeBlock, CodeExecutor, extract_markdown_code_blocks
from autogen_core.components.models import UserMessage

from .._base_chat_agent import BaseChatAgent, ChatMessage


class CodeExecutorAgent(BaseChatAgent):
    """An agent that executes code snippets and report the results."""

    DESCRIPTION = "A computer terminal that performs no other action than running Python scripts (provided to it quoted in ```python code blocks), or sh shell scripts (provided to it quoted in ```sh code blocks)."

    def __init__(self, name: str, code_executor: CodeExecutor):
        """Initialize the agent with a code executor."""
        super().__init__(name=name, description=self.DESCRIPTION)
        self._code_executor = code_executor

    async def on_messages(self, messages: Sequence[ChatMessage], cancellation_token: CancellationToken) -> ChatMessage:
        # Extract code blocks from the messages.
        code_blocks: List[CodeBlock] = []
        for msg in messages:
            if isinstance(msg.content, UserMessage) and isinstance(msg.content.content, str):
                code_blocks.extend(extract_markdown_code_blocks(msg.content.content))
        if code_blocks:
            # Execute the code blocks.
            result = await self._code_executor.execute_code_blocks(code_blocks, cancellation_token=cancellation_token)
            return ChatMessage(content=UserMessage(content=result.output, source=self.name), request_pause=False)
        else:
            return ChatMessage(
                content=UserMessage(content="No code blocks found in the thread.", source=self.name),
                request_pause=False,
            )

File: dotnet/src/Microsoft.AutoGen/AgentChat/Agents/CodeExecutorAgent.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// CodeExecutorAgent.cs

using System.Text.RegularExpressions;
using Microsoft.AutoGen.Abstractions;
using Microsoft.AutoGen.AgentChat.Abstractions;

using TextMessage = Microsoft.AutoGen.AgentChat.Abstractions.TextMessage;

namespace Microsoft.AutoGen.AgentChat.Agents;

// TODO: Should this live in one of the Core packages (similar to .components in python?)
public static partial class MarkdownExtensions
{
    private static readonly Regex codeBlockPattern = new Regex(@"```(?:\s*(?<language>[\w\+\-]+))?\r?\n(?<code>[\s\S]*)```", RegexOptions.Compiled | RegexOptions.Multiline);
    public static IEnumerable<CodeBlock> ExtractCodeBlocks(this string markdownText)
    {
        return codeBlockPattern.Matches(markdownText)
                               .Select((match) =>
                                   new CodeBlock
                                   {
                                       Code = match.Groups["code"].Value.TrimEnd(),
                                       Language = match.Groups["language"].Value.Trim()
                                   });
    }
}

public class CodeExecutorAgent : ChatAgentBase
{
    private const string DefaultDescription = "A computer terminal that performs no other action than running scripts (provided to it quoted in ```<language>> code blocks).";
    public CodeExecutorAgent(string name, ICodeExecutor codeExecutor, string description = DefaultDescription) : base(name, description)
    {
        CodeExecutor = codeExecutor;
    }

    public ICodeExecutor CodeExecutor { get; }

    public override IEnumerable<Type> ProducedMessageTypes => [typeof(TextMessage)];

    public override async ValueTask<Response> HandleAsync(IEnumerable<ChatMessage> messages, CancellationToken cancellationToken)
    {
        var codeBlocks = messages.OfType<TextMessage>()
                                                    .SelectMany((textMessage) => textMessage.Content.ExtractCodeBlocks());

        var result = "No code blocks found.";
        if (codeBlocks.Any())
        {
            var codeResult = await CodeExecutor.ExecuteCodeBlocksAsync(codeBlocks, cancellationToken);
            result = codeResult.Output;
        }

        return new Response { Message = new TextMessage { Content = result, Source = Name } };
    }

    public override ValueTask ResetAsync(CancellationToken cancellationToken)
    {
        return ValueTask.CompletedTask;
    }
}
